### ISSUE TITLE
cluster continues to scale up when SGE resource requirements not met

### ISSUE THREAD

**Issue by tkuzuhara** [aws/aws-parallelcluster#541]
In using cfncluster with SGE, we have cases where we define SGE consumable resources to track software licenses.  These resources are added to the SGE configuration through a custom post-install script.

If a job cannot run due to lack of resources (say lack of a license), that job will sit in the `qw` state until a license becomes available.  Unfortunately it appears the cloudwatch metric for pending jobs still counts these jobs, and thus cfncluster appears to continuously try to scale the cluster up.  Of course since the real bottleneck is the number of available licenses, scaling the cluster provides no benefit.  In cases where there is an extended backlog of licenses, this can cause the cluster to scale up to the max unnecessarily.

Is there a way to further qualify the number of pending jobs metric with SGE to ignore jobs that are pending due to lack of consumable resource?

**Comment by enrico-usai** [comment]
Hi @tkuzuhara ,
this bug has been resolved in the upcoming 1.6.0 release, thanks to an hard scaling up/down refactoring:
https://github.com/awslabs/cfncluster/blob/develop/CHANGELOG.rst#160

If you are willing to test the release candindate you can install it with:
`pip install -Iv cfncluster==1.6.0rc1`

Please keep us updated if you have the opportunity to test the rc1 or please stay tuned for the upcoming 1.6.0 release.

**Comment by tkuzuhara** [comment]
Hi, we upgraded to v1.6.1 but still have this issue.

In case I wasn't clear, this has to do specifically with sge consumable resources.  cfncluster as far as I can tell only tracks whether a job is in the `qw` state, but doesn't know why.  In cases of a consumable resource, a job may appear in the `qw` state, but scaling up compute nodes will not help the situation because the underlying resource is unaffected by spinning up new nodes.

An example of using consumable resources is for tracking consumable software licenses.  You might have only 10 licenses of a product across the company.  You want up to 10 jobs on the cluster to be running this software, and any further requests should stay queued until one of the 10 jobs consuming the license are freed.

Here are some resources on setting up these consumable resources:
http://gridscheduler.sourceforge.net/howto/consumable.html
https://docs.oracle.com/cd/E19957-01/820-0698/chp8-1501/index.html

For convenience, sometimes people might batch a large number of jobs that is reliant on a small pool of licenses.  In doing so, you end up with a large number of jobs in `qw` because of the license crunch.  I think the current implementation of jobwatcher does not know that the large number of jobs are waiting on a resource, rather than waiting for a slot, and thus will spin up machines in this case, up to the size of jobs in `qw`.  These jobs still won't be able to run so the machines that were spun up will sit idle.

Here is an example.  I've set up a test cluster that is pretty much the default hello-world cluster (https://cfncluster.readthedocs.io/en/latest/hello_world.html).  Each compute node can take at most one job.  

Start out with two nodes in the cluster:
```
[centos@ip-172-31-128-41 ~]$ qhost
HOSTNAME                ARCH         NCPU NSOC NCOR NTHR  LOAD  MEMTOT  MEMUSE  SWAPTO  SWAPUS
----------------------------------------------------------------------------------------------
global                  -               -    -    -    -     -       -       -       -       -
ip-172-31-128-187       lx-amd64        1    1    1    1  0.16  990.1M  311.3M     0.0     0.0
ip-172-31-128-190       lx-amd64        1    1    1    1  0.01  990.1M  277.3M     0.0     0.0
```

Now launch asking for 6 slots.  Note that this is impossible for this cluster to satisfy, since each machine can take at most 1 slot.  I'm using slots here as a proxy for licenses.  The point is that adding more machines will not help.
```
[centos@ip-172-31-128-41 ~]$ qsub -pe smp 6 hellojob.sh
Your job 3 ("hellojob.sh") has been submitted
```

This job will be in queue wait indefinitely:
```
[centos@ip-172-31-128-41 ~]$ qstat
job-ID  prior   name       user         state submit/start at     queue                          slots ja-task-ID 
-----------------------------------------------------------------------------------------------------------------
      3 0.55500 hellojob.s centos       qw    11/02/2018 00:09:31                                    6       
```

However jobwatcher will still think that launching new hosts will help.  It thinks that 6 hosts will satisfy the needs of this job.  As a result the cluster scales to 6.  The job of course stays in `qw`:
```
[centos@ip-172-31-128-41 ~]$ qhost
HOSTNAME                ARCH         NCPU NSOC NCOR NTHR  LOAD  MEMTOT  MEMUSE  SWAPTO  SWAPUS
----------------------------------------------------------------------------------------------
global                  -               -    -    -    -     -       -       -       -       -
ip-172-31-128-137       lx-amd64        1    1    1    1  0.36  990.1M  295.0M     0.0     0.0
ip-172-31-128-141       lx-amd64        1    1    1    1  0.43  990.1M  292.0M     0.0     0.0
ip-172-31-128-162       lx-amd64        1    1    1    1  0.32  990.1M  296.2M     0.0     0.0
ip-172-31-128-187       lx-amd64        1    1    1    1  0.12  990.1M  357.1M     0.0     0.0
ip-172-31-128-190       lx-amd64        1    1    1    1  0.23  990.1M  317.4M     0.0     0.0
ip-172-31-128-50        lx-amd64        1    1    1    1  0.34  990.1M  295.3M     0.0     0.0
[centos@ip-172-31-128-41 ~]$ qstat
job-ID  prior   name       user         state submit/start at     queue                          slots ja-task-ID 
-----------------------------------------------------------------------------------------------------------------
      3 0.55500 hellojob.s centos       qw    11/02/2018 00:09:31                                    6   
```

Here is the relevant portion of `/var/log/jobwatcher`:
```
2018-11-02 00:09:37,854 INFO [jobwatcher:main] 6 jobs pending; 0 jobs running
2018-11-02 00:09:38,052 INFO [jobwatcher:main] min/desired/max 2/2/10
2018-11-02 00:09:38,052 INFO [jobwatcher:main] Nodes requested 6, Nodes running 0
2018-11-02 00:09:38,052 INFO [jobwatcher:main] Setting desired to 6 nodes, requesting 4 more nodes from asg.
2018-11-02 00:10:38,342 INFO [jobwatcher:get_vcpus_from_pricing_file] Instance t2.micro has 1 vcpus.
2018-11-02 00:10:38,342 INFO [jobwatcher:get_instance_properties] Instance t2.micro will use number of vcpus as slots based on configuration.
2018-11-02 00:10:38,365 INFO [jobwatcher:main] 6 jobs pending; 0 jobs running
2018-11-02 00:10:38,489 INFO [jobwatcher:main] min/desired/max 2/6/10
2018-11-02 00:10:38,489 INFO [jobwatcher:main] Nodes requested 6, Nodes running 0
2018-11-02 00:10:38,489 INFO [jobwatcher:main] 6 nodes desired 6 nodes in asg. Noop
```

I'm not sure if there is an easy way to find out *why* a job is in `qw`.  (Maybe `qalter -w p ...`?)  Clearly a job waiting on a consumable resource that won't be satisfied by adding new nodes should not cause cfncluster to scale up.

One possible heuristic would be to stop scaling if there are N slots free on the cluster, where N is a configurable number to the job watcher.  If there are N slots free on the cluster, do not scale up, regardless of the number of jobs waiting.  The logic here being that if you've added N slots and jobs are *still* waiting, then adding more slots will not rectify the situation.  This will help avoid cases where the scaling continues indefinitely.  Maybe there are better ways to handle this...

**Comment by mauri-melato** [comment]
Hi tkuzuhara, 
your catch is correct, today scale up decisions are only made on the base of the computational  resources required, i.e. the number of slots required to run all the jobs in the queue.
The new scale up asks for the needed resources at once, 4 new nodes in your test, and it doesn't continue to scale up indefinitely as you could see in your last log message. Once there are the proper number of computational resources available to satisfy the pending jobs it stops, so you're not scaling up to maximum of 10 nodes.

Now the usage scenario you report, where consumable resources are defined for tracking software licenses makes perfectly sense, and it's a very good feedback for a request for enhancement I'm going to register.
The how to finely distinguish pending jobs conditions/reasons needs be investigated on a scheduler basis.

**Comment by nstornetta** [comment]
Because we have announced that we will be deprecating support for SGE in the near-future (see: https://github.com/aws/aws-parallelcluster/wiki/Deprecation-of-SGE-and-Torque-in-ParallelCluster), we will not be performing additional enhancements specific to SGE.

I am going to close this issue. If you would like to request a similar enhancement for one of our other supported schedulers (Slurm or AWS Batch), please feel free to create a new issue.
